node/silkworm/rpc/call.hpp

```diff
@@ -0,0 +1,222 @@
+
+namespace silkworm::rpc {
+
+//! This represents the generic gRPC call composed by a sequence of bidirectional operations.
+class BaseRpc {
+  public:
+    static int32_t instance_count() { return instance_count_; }
+
+    BaseRpc() : op_count_(0), done_(false) {
+        instance_count_++;
+        SILK_TRACE << "BaseRpc::BaseRpc [" << this << "] instances: " << instance_count_;
+    }
+
+    virtual ~BaseRpc() {
+        instance_count_--;
+        SILK_TRACE << "BaseRpc::~BaseRpc [" << this << "] instances: " << instance_count_;
+    }
+
+    /// Tag processor for the DONE event in this RPC coming from gRPC framework.
+    void process_done(bool ok) {
+        SILK_TRACE << "BaseRpc::on_done START ok: " << ok << " done_: " << done_ << " op_count_: " << op_count_;
+        done_ = true;
+        if (op_count_ == 0) {
+            cleanup();
+        }
+        SILK_TRACE << "BaseRpc::on_done END ok: " << ok << " op_count_: " << op_count_;
+    }
+
+  protected:
+    /// Hook to signal this RPC is *really* done: each subclass shall override to implement its own cleanup.
+    virtual void cleanup() = 0;
+
+    //! The 4 different types of bidirectional operations composing a RPC.
+    /// Some operations may occur multiple times during a single RPC lifetime.
+    enum class OperationType {
+        kRequest,
+        kRead,
+        kWrite,
+        kFinish
+    };
+
+    /// Callback to handle the start of the specified async operation.
+    void handle_started(OperationType opType) {
+        ++op_count_;
+        SILK_TRACE << "BaseRpc::handle_started opType: " << static_cast<int>(opType) << " op_count_: " << op_count_;
+    }
+
+    /// Callback to handle the end of the specified async operation.
+    /// \return true if the RPC processing should keep going, false otherwise
+    bool handle_completed(OperationType opType) {
+        --op_count_;
+        SILK_TRACE << "BaseRpc::handle_completed opType: " << static_cast<int>(opType) << " op_count_: " << op_count_;
+
+        if (op_count_ == 0 && done_) {
+            // No async operations pending and gRPC library already notified that this RPC is done => cleanup.
+            cleanup();
+            return false;
+        }
+
+        return true;
+    }
+
+    /// Used to access the options and current status of the RPC.
+    grpc::ServerContext context_;
+
+    /// Keep track of the total outstanding RPC calls (intentionally signed to spot underflows).
+    inline static std::atomic_int64_t instance_count_ = 0;
+
+  private:
+    /// This counts the number of pending operations in this RPC.
+    /// \warning It is used to detect when the RPC is really done in some corner cases (e.g. client abruptly exits).
+    uint32_t op_count_;
+
+    /// This flag indicates if DONE tag processor has been called or not.
+    /// \warning It is used to handle some corner cases when the completion queue dispatches tags in reverse order
+    /// (e.g. client abruptly exits).
+    bool done_;
```

Comment:
My preference is to use designated initializers at the declaration. Thus one can be sure that the variable can't be uninitialized without going through all constructors.